packages/grafana-data/src/field/fieldOverrides.ts

```diff
@@ -0,0 +1,213 @@
+import set from 'lodash/set';
+import {
+  DynamicConfigValue,
+  FieldConfigSource,
+  FieldConfig,
+  InterpolateFunction,
+  GrafanaTheme,
+  DataFrame,
+  Field,
+  FieldType,
+} from '../types';
+import { fieldMatchers, ReducerID, reduceField } from '../transformations';
+import { FieldMatcher } from '../types/transformations';
+import isNumber from 'lodash/isNumber';
+import toNumber from 'lodash/toNumber';
+import { getDisplayProcessor } from './displayProcessor';
+
+interface OverrideProps {
+  match: FieldMatcher;
+  properties: DynamicConfigValue[];
+}
+
+interface GlobalMinMax {
+  min: number;
+  max: number;
+}
+
+export function findNumericFieldMinMax(data: DataFrame[]): GlobalMinMax {
+  let min = Number.MAX_VALUE;
+  let max = Number.MIN_VALUE;
+
+  const reducers = [ReducerID.min, ReducerID.max];
+  for (const frame of data) {
+    for (const field of frame.fields) {
+      if (field.type === FieldType.number) {
+        const stats = reduceField({ field, reducers });
+        if (stats[ReducerID.min] < min) {
+          min = stats[ReducerID.min];
+        }
+        if (stats[ReducerID.max] > max) {
+          max = stats[ReducerID.max];
+        }
+      }
+    }
+  }
+
+  return { min, max };
+}
+
+/**
+ * Return a copy of the DataFrame with all rules applied
+ */
+export function applyFieldOverrides(
+  data: DataFrame[],
+  source: FieldConfigSource,
+  replaceVariables: InterpolateFunction,
+  theme: GrafanaTheme,
+  isUtc?: boolean
+): DataFrame[] {
+  if (!source) {
+    return data;
+  }
+  let range: GlobalMinMax | undefined = undefined;
+
+  // Prepare the Matchers
+  const override: OverrideProps[] = [];
+  if (source.overrides) {
+    for (const rule of source.overrides) {
+      const info = fieldMatchers.get(rule.matcher.id);
+      if (info) {
+        override.push({
+          match: info.get(rule.matcher),
+          properties: rule.properties,
+        });
+      }
+    }
+  }
+
+  return data.map((frame, index) => {
+    let name = frame.name;
+    if (!name) {
+      name = `Series[${index}]`;
+    }
+
+    const fields = frame.fields.map(field => {
+      // Config is mutable within this scope
+      const config: FieldConfig = { ...field.config } || {};
+      if (field.type === FieldType.number) {
+        setFieldConfigDefaults(config, source.defaults);
+      }
+
+      // Find any matching rules and then override
+      for (const rule of override) {
+        if (rule.match(field)) {
+          for (const prop of rule.properties) {
+            setDynamicConfigValue(config, {
```

Comment:
Regarding the other comment about mutating the config- I feel I see your point now. Makes this code easier without the need for reassignments

Reply:
yes -- reassignments got pretty ugly